File: web/src/constants/permissions.ts
export const ROLE_SUPER_ADMIN = 'super_admin';

export const PERMISSIONS = {
  dashboardView: 'dashboard:view',
  operationLogView: 'operation_log:view',
  portalAccess: 'portal:access',
  portalManage: 'portal:manage',
  announcementManage: 'announcement:manage',
  searchKeywordManage: 'search_keyword:manage',
  featuredResourceManage: 'featured_resource:manage',
  contentAccess: 'content:access',
  templateManage: 'template:manage',
  applicationCaseManage: 'application_case:manage',
  documentCategoryManage: 'document_category:manage',
  documentManage: 'document:manage',
  faqManage: 'faq:manage',
  videoManage: 'video:manage',
  agreementManage: 'agreement_template:manage',
  privacyManage: 'privacy_option:manage',
  filterOptionManage: 'filter_option:manage',
  modelRecommendTagManage: 'model_recommend_tag:manage',
  reviewAccess: 'review:access',
  reviewManage: 'review:manage',
  datasetAccessAccess: 'dataset_access:access',
  datasetAccessReview: 'dataset_access:review',
  verificationAccess: 'verification:access',
  verificationReview: 'verification:review',
  communityAccess: 'community:access',
  communityContentModerate: 'community_content:moderate',
  conversationModerate: 'conversation:moderate',
  workspaceModerate: 'workspace:moderate',
  wikiManage: 'wiki:manage',
  rewardAccess: 'reward:access',
  rewardManage: 'reward:manage',
} as const;

const allPermissions = Object.values(PERMISSIONS);

export const rolePermissionMatrix: Record<string, string[]> = {
  user: [],
  developer: [],
  operator: [
    PERMISSIONS.dashboardView,
    PERMISSIONS.operationLogView,
    PERMISSIONS.portalAccess,
    PERMISSIONS.portalManage,
    PERMISSIONS.announcementManage,
    PERMISSIONS.searchKeywordManage,
    PERMISSIONS.featuredResourceManage,
    PERMISSIONS.contentAccess,
    PERMISSIONS.templateManage,
    PERMISSIONS.applicationCaseManage,
    PERMISSIONS.documentCategoryManage,
    PERMISSIONS.documentManage,
    PERMISSIONS.faqManage,
    PERMISSIONS.videoManage,
    PERMISSIONS.agreementManage,
    PERMISSIONS.privacyManage,
    PERMISSIONS.filterOptionManage,
    PERMISSIONS.modelRecommendTagManage,
    PERMISSIONS.communityAccess,
    PERMISSIONS.communityContentModerate,
    PERMISSIONS.conversationModerate,
    PERMISSIONS.workspaceModerate,
  ],
  reviewer: [
    PERMISSIONS.dashboardView,
    PERMISSIONS.operationLogView,
    PERMISSIONS.reviewAccess,
    PERMISSIONS.reviewManage,
    PERMISSIONS.datasetAccessAccess,
    PERMISSIONS.datasetAccessReview,
    PERMISSIONS.verificationAccess,
    PERMISSIONS.verificationReview,
    PERMISSIONS.communityAccess,
    PERMISSIONS.communityContentModerate,
    PERMISSIONS.conversationModerate,
    PERMISSIONS.workspaceModerate,
  ],
  admin: [...allPermissions],
  [ROLE_SUPER_ADMIN]: [...allPermissions],
};

export function permissionsByRole(role?: string) {
  return rolePermissionMatrix[role ?? ''] ?? [];
}

export function resolvePermissions(user?: { role?: string; permissions?: string[] | null }) {
  if (user?.permissions?.length) {
    return [...user.permissions];
  }
  return permissionsByRole(user?.role);
}

File: server/internal/model/permissions.go
package model

import "slices"

const (
	RoleSuperAdmin = "super_admin"
)

const (
	PermissionDashboardView            = "dashboard:view"
	PermissionOperationLogView         = "operation_log:view"
	PermissionPortalAccess             = "portal:access"
	PermissionPortalManage             = "portal:manage"
	PermissionAnnouncementManage       = "announcement:manage"
	PermissionSearchKeywordManage      = "search_keyword:manage"
	PermissionFeaturedResourceManage   = "featured_resource:manage"
	PermissionContentAccess            = "content:access"
	PermissionTemplateManage           = "template:manage"
	PermissionApplicationCaseManage    = "application_case:manage"
	PermissionDocumentCategoryManage   = "document_category:manage"
	PermissionDocumentManage           = "document:manage"
	PermissionFAQManage                = "faq:manage"
	PermissionVideoManage              = "video:manage"
	PermissionAgreementManage          = "agreement_template:manage"
	PermissionPrivacyManage            = "privacy_option:manage"
	PermissionFilterOptionManage       = "filter_option:manage"
	PermissionModelRecommendTagManage  = "model_recommend_tag:manage"
	PermissionReviewAccess             = "review:access"
	PermissionReviewManage             = "review:manage"
	PermissionDatasetAccessAccess      = "dataset_access:access"
	PermissionDatasetAccessReview      = "dataset_access:review"
	PermissionVerificationAccess       = "verification:access"
	PermissionVerificationReview       = "verification:review"
	PermissionCommunityAccess          = "community:access"
	PermissionCommunityContentModerate = "community_content:moderate"
	PermissionConversationModerate     = "conversation:moderate"
	PermissionWorkspaceModerate        = "workspace:moderate"
	PermissionWikiManage               = "wiki:manage"
	PermissionRewardAccess             = "reward:access"
	PermissionRewardManage             = "reward:manage"
)

var allPermissions = []string{
	PermissionDashboardView,
	PermissionOperationLogView,
	PermissionPortalAccess,
	PermissionPortalManage,
	PermissionAnnouncementManage,
	PermissionSearchKeywordManage,
	PermissionFeaturedResourceManage,
	PermissionContentAccess,
	PermissionTemplateManage,
	PermissionApplicationCaseManage,
	PermissionDocumentCategoryManage,
	PermissionDocumentManage,
	PermissionFAQManage,
	PermissionVideoManage,
	PermissionAgreementManage,
	PermissionPrivacyManage,
	PermissionFilterOptionManage,
	PermissionModelRecommendTagManage,
	PermissionReviewAccess,
	PermissionReviewManage,
	PermissionDatasetAccessAccess,
	PermissionDatasetAccessReview,
	PermissionVerificationAccess,
	PermissionVerificationReview,
	PermissionCommunityAccess,
	PermissionCommunityContentModerate,
	PermissionConversationModerate,
	PermissionWorkspaceModerate,
	PermissionWikiManage,
	PermissionRewardAccess,
	PermissionRewardManage,
}

var rolePermissionMatrix = map[string][]string{
	RoleUser:      {},
	RoleDeveloper: {},
	RoleOperator: {
		PermissionDashboardView,
		PermissionOperationLogView,
		PermissionPortalAccess,
		PermissionPortalManage,
		PermissionAnnouncementManage,
		PermissionSearchKeywordManage,
		PermissionFeaturedResourceManage,
		PermissionContentAccess,
		PermissionTemplateManage,
		PermissionApplicationCaseManage,
		PermissionDocumentCategoryManage,
		PermissionDocumentManage,
		PermissionFAQManage,
		PermissionVideoManage,
		PermissionAgreementManage,
		PermissionPrivacyManage,
		PermissionFilterOptionManage,
		PermissionModelRecommendTagManage,
		PermissionCommunityAccess,
		PermissionCommunityContentModerate,
		PermissionConversationModerate,
		PermissionWorkspaceModerate,
	},
	RoleReviewer: {
		PermissionDashboardView,
		PermissionOperationLogView,
		PermissionReviewAccess,
		PermissionReviewManage,
		PermissionDatasetAccessAccess,
		PermissionDatasetAccessReview,
		PermissionVerificationAccess,
		PermissionVerificationReview,
		PermissionCommunityAccess,
		PermissionCommunityContentModerate,
		PermissionConversationModerate,
		PermissionWorkspaceModerate,
	},
	RoleAdmin:      append([]string{}, allPermissions...),
	RoleSuperAdmin: append([]string{}, allPermissions...),
}

func RolePermissions(role string) []string {
	permissions, ok := rolePermissionMatrix[role]
	if !ok {
		return []string{}
	}
	return append([]string{}, permissions...)
}

func HasPermission(role, permission string) bool {
	if permission == "" {
		return true
	}
	return slices.Contains(RolePermissions(role), permission)
}

func HasAllPermissions(role string, permissions ...string) bool {
	for _, permission := range permissions {
		if !HasPermission(role, permission) {
			return false
		}
	}
	return true
}

func IsAdminRole(role string) bool {
	return role == RoleAdmin || role == RoleSuperAdmin
}

func HasBackofficeAccess(role string) bool {
	return len(RolePermissions(role)) > 0
}
